securedrop_client/gui/widgets.py

```diff
@@ -789,52 +789,57 @@ def update(self, sources: List[Source]) -> List[str]:
-            list_widget = self.itemWidget(list_item)
+        sources_to_delete = [self.source_items[uuid] for uuid in self.source_items
+                             if uuid not in sources_to_update]
+        for source_item in sources_to_delete:
+            if source_item.isSelected():
+                self.setCurrentItem(None)
 
-            if list_widget and list_widget.source_uuid not in source_uuids:
-                if list_item.isSelected():
-                    self.setCurrentItem(None)
+            source_widget = self.itemWidget(source_item)
+            self.takeItem(self.row(source_item))
+            source_widget.deleteLater()
+            if source_widget.source_uuid in self.source_items:
+                del self.source_items[source_widget.source_uuid]
 
-                try:
-                    del self.source_widgets[list_widget.source_uuid]
-                except KeyError:
-                    pass
+            deleted_uuids.append(source_widget.source_uuid)
 
-                self.takeItem(i)
-                deleted_uuids.append(list_widget.source_uuid)
-                list_widget.deleteLater()
+        # Update the remaining widgets
+        for i in range(self.count()):
+            source_widget = self.itemWidget(self.item(i))
 
-        # Create new widgets for new sources
-        widget_uuids = [self.itemWidget(self.item(i)).source_uuid for i in range(self.count())]
-        for source in sources:
-            if source.uuid in widget_uuids:
-                try:
-                    self.source_widgets[source.uuid].update()
-                except sqlalchemy.exc.InvalidRequestError as e:
-                    logger.error(
-                        "Could not update SourceWidget for source %s; deleting it. Error was: %s",
-                        source.uuid,
-                        e
-                    )
-                    deleted_uuids.append(source.uuid)
-                    self.source_widgets[source.uuid].deleteLater()
-                    del self.source_widgets[list_widget.source_uuid]
-            else:
-                new_source = SourceWidget(self.controller, source)
-                self.source_widgets[source.uuid] = new_source
+            if not source_widget:
+                continue
+
+            source_widget.update()
 
-                list_item = SourceListWidgetItem()
-                self.insertItem(0, list_item)
-                list_item.setSizeHint(new_source.sizeHint())
-                self.setItemWidget(list_item, new_source)
+        # Create new source widgets for new sources
+        for source in sources_to_add:
+            source_widget = SourceWidget(self.controller, source)
+            source_item = SourceListWidgetItem(self)
+            source_item.setSizeHint(source_widget.sizeHint())
+            self.insertItem(0, source_item)
+            self.setItemWidget(source_item, source_widget)
+            self.source_items[source.uuid] = source_item
```

Comment:
(re: #906) Since the source database object can be deleted in the sync thread I think we'll want try/except here also to make this `uuid` attribute access safe

Reply:
yup! you're right, or we could do what i originally intended to do here: make sources_to_add a safe-to-access type dict[uuid, Source]